src/agent/onefuzz-task/src/tasks/coverage/generic.rs

```diff
@@ -457,49 +457,78 @@ impl<'a> TaskContext<'a> {
-        let json = BinaryCoverageJson::V1(BinaryCoverageJsonV1::from(binary));
-        let text = serde_json::to_string(&json).context("serializing binary coverage")?;
-        let path = self.config.coverage.local_path.join(COVERAGE_FILE);
-        fs::write(&path, &text)
-            .await
-            .with_context(|| format!("writing coverage to {}", path.display()))?;
-
-        // JSON source coverage.
-        let source = self.source_coverage().await?;
-        let json = SourceCoverageJson::V1(SourceCoverageJsonV1::from(source.clone()));
-        let text = serde_json::to_string(&json).context("serializing source coverage")?;
-        let path = self.config.coverage.local_path.join(SOURCE_COVERAGE_FILE);
-        fs::write(&path, &text)
-            .await
-            .with_context(|| format!("writing source coverage to {}", path.display()))?;
+    pub async fn save_coverage(
+        coverage: &BinaryCoverage,
+        source_allowlist: &AllowList,
+        binary_coverage_path: &Path,
+        source_coverage_path: &Path,
+        copbertura_file_path: &Path,
+    ) -> Result<()> {
+        Self::save_binary_coverage(coverage, binary_coverage_path)?;
+
+        let source = binary_to_source_coverage(coverage, source_allowlist.clone())?;
+
+        Self::save_source_coverage(&source, source_coverage_path).await?;
+
+        Self::save_cobertura_xml(&source, copbertura_file_path).await?;
+        Ok(())
+    }
 
-        // Cobertura XML source coverage.
-        let cobertura = CoberturaCoverage::from(source.clone());
-        let text = cobertura.to_string()?;
-        let path = self
+    pub async fn save_and_sync_coverage(&self) -> Result<()> {
+        let copbertura_file_path = self
             .config
             .coverage
             .local_path
             .join(COBERTURA_COVERAGE_FILE);
-        fs::write(&path, &text)
-            .await
-            .with_context(|| format!("writing cobertura source coverage to {}", path.display()))?;
 
+        let source_coverage_path = self.config.coverage.local_path.join(SOURCE_COVERAGE_FILE);
+        let binary_coverage_path = self.config.coverage.local_path.join(COVERAGE_FILE);
+
+        Self::save_coverage(
+            &self.coverage,
+            &self.source_allowlist,
+            &binary_coverage_path,
+            &source_coverage_path,
+            &copbertura_file_path,
+        )
+        .await?;
         self.config.coverage.sync_push().await?;
+        Ok(())
+    }
 
+    async fn save_cobertura_xml(source: &SourceCoverage, path: &Path) -> Result<(), anyhow::Error> {
+        let cobertura = CoberturaCoverage::from(source);
+        let cobertura_coverage_file = std::fs::File::create(path)
+            .with_context(|| format!("creating cobertura coverage file {}", path.display()))?;
+        let cobertura_coverage_file_writer = std::io::BufWriter::new(cobertura_coverage_file);
+        cobertura
+            .write_xml(cobertura_coverage_file_writer)
```

Comment:
It might be worthwhile (as a forever backlog feature) to add `AsyncWrite` support for `write_xml`, that way we can use tokio's `BufWriter`. I'm not sure how much time we spend blocking the thread to write these coverage files but I guess if they get big enough there's some value there. Especially since we write the file every time we merge the coverage which I think is on every input processed?

Reply:
#3419